Tuple parameter type.

ezzatron opened the pull request:
This PR introduces a `tuple` type to the types appendix, inspired by [Typhax's tuple type syntax](https://github.com/eloquent/typhax#tuple).

Tuples are useful constructs that are formalized in many languages. They are essentially fixed-size arrays with specific types for each element. They are very useful in conjunction with the [list()](http://php.net/list) language construct:

``` php
$tuple = ['a', 1, true];             // type definition is tuple<string,int,bool>
list($string, $int, $bool) = $tuple; // unpacking the tuple using list()
```

As a side note, [Typhax](https://github.com/eloquent/typhax) is a project I wrote to define a more expressive type syntax for parameter documentation. It's probably only used by devs at the company I work for, and a few friends, but it's been very successful nonetheless.

 I'd like to open a few PRs to introduce ideas from Typhax, and generate some discussion. Hopefully I'm not too late to the game!

Fleshgrinder commented:
:+1: from my side, useful scalar hint for recurring data structures that would be overhead for creating dedicated classes.

sun commented:
Any particular reason for not naming this type `list` instead of `tuple`?

```diff
 /**
- * @return tuple<int,int,int>
+ * @return list<int,int,int>
  */
```

`list` is a [reserved keyword](http://php.net/manual/en/reserved.keywords.php) in PHP, so it cannot appear as a possible type anyway.

Fleshgrinder commented:
Naming it list might seem obvious to you as a PHP developer, however, the name _tuple_ has a mathematical background and is therefore more descriptive than anything else could ever be.

https://en.wikipedia.org/wiki/Tuple

ezzatron commented:
Personally, it doesn't bother me too much what the keyword is, but given that most of the existing keywords could be defined as classes anyway, and this note that already exists in the spec:

> Most of these keywords are allowed as class names in PHP and as such are hard to distinguish from real classes. As such the keywords MUST be lowercase, as most class names start with an uppercase first character, and you SHOULD NOT use classes with these names in your code.
>
> There are more reasons to not name classes with the names of these keywords but that falls beyond the scope of this specification.

I'd probably lean more towards using `tuple`, for clarity. I'd assume that if you really wanted to call your class `tuple` (lowercase), then you could still document it using a qualified class name:

```
@returns \tuple
@returns \my\ns\tuple
```

ezzatron commented:
I just updated this PR, because I realized I screwed up the ABNF, *then* I realized that tuple wouldn't actually be a keyword, and needs its own ABNF rule.

So I moved the description out of the "Keywords" section, and into its own section, just under arrays and collections, gave it a little re-write, and fixed the ABNF properly.

For some reason GitHub isn't showing that I've updated this PR though. Probably because I squashed the commits?

Fleshgrinder commented:
Just tested it in PhpStorm, it does not distinguish a class name by casing. I would say this is a bug.

dennisbirkholz commented:
> Just tested it in PhpStorm, it does not distinguish a class name by casing. I would say this is a bug.

Class names in PHP are not case sensitive so PhpStorm behaves as it should.

ezzatron commented:
Agreed, that sounds like correct behaviour. The PSR-5 draft even states that care must be taken not to name classes that will clash with type keywords. To get around this, you should be able to use an alias; e.g. `use tuple as MyTuple;`.

CC007 commented:
Is there a reason why this branch is still open after 2 years, but is not merged yet?

ashnazg commented:
This repo is outdated... please get a new fork from https://github.com/php-fig/fig-standards for future contributions.

This *does* still need to be discussed in the FIG repo, so please open a new PR there.

ashnazg commented:
Rather than working against FIG repo, I have now updated this phpDoc fork, so we can resume working here.

Please update your fork of this repo, and update this PR to resolve its conflicts.

mindplay-dk commented:
Is this supported by more than one library?

I checked, and neither phpstan, phan, psalm or storm seem to support tuples.

This is really useful, however. Not that I'm hoping anyone is still out there writing new code that makes heavy use of tuples, but there are a lot of functions built-into PHP that could benefit from this - and I don't see the discussion about the details of this having the potential to get too terribly out of hand?

So I wouldn't be opposed as such, though I don't want to invite scope creep, so.

I guess I'm neutral on this one.

ashnazg commented:
If someone feels strongly about keeping this change, please bring it up on the FIG mailing list for discussion -- https://groups.google.com/forum/#!forum/php-fig
